Feature: "Remove all industries" button in scenario editor

Kuhnovic opened the pull request:
## Motivation / Problem

Many scenarios found in on the online content server come with industries. This is unfortunate when you want to play that particular map with for instance the FIRS industry set, or if you want to start with a clean slate for a different reason. The current solution is to go over the entire map and remove each industry, which is tedious.

## Description

This feature is only available in the scenario editor. It removes all industries. It also turns removes residual farmland and replaces it with grass.

![wAc0puysLU](https://user-images.githubusercontent.com/68320206/104109093-94e77400-52ca-11eb-8be1-3500319c8ff8.gif)

## Limitations


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

Kuhnovic committed:
Feature: Remove all industries button in scenario editor

glx22 commented on src/industry_gui.cpp:
```diff
@@ -332,6 +337,8 @@ class BuildIndustryWindow : public Window {
 	{
 		this->SetWidgetDisabledState(WID_DPI_FUND_WIDGET, this->selected_type != INVALID_INDUSTRYTYPE && !this->enabled[this->selected_index]);
 		this->SetWidgetDisabledState(WID_DPI_DISPLAY_WIDGET, this->selected_type == INVALID_INDUSTRYTYPE && this->enabled[this->selected_index]);
+
```
Useless empty lines

glx22 commented on src/industry_gui.cpp:
```diff
@@ -578,9 +592,35 @@ class BuildIndustryWindow : public Window {
 
+	static void AskRemoveAllIndustriesCallback(Window *w, bool confirmed)
+	{
+		if (confirmed) {
```
Better do `if (!confirmed) return;`, then you can reduce indentation level

glx22 commented on src/industry_gui.cpp:
```diff
@@ -392,6 +399,13 @@ class BuildIndustryWindow : public Window {
 		this->SetButtons();
+
+		/* Show scenario editor tools in editor */
+		auto *se_tools = this->GetWidget<NWidgetStacked>(WID_DPI_SCENARIO_EDITOR_PANE);
```
The variable is used only inside `if` block, so better not define it outside.

glx22 commented on src/industry_gui.cpp:
```diff
@@ -578,9 +592,35 @@ class BuildIndustryWindow : public Window {
+			for (int i = 0; i < numberOfIndustries; i++) {
+				auto* industry = Industry::Get(i);
+				if (industry) delete industry;
+			}
```
```suggestion
			for (Industry *i : Industry::Iterate()) delete i;
```

Kuhnovic replied:
I must have overlooked the Iterate method, this is much cleaner indeed, thanks!

glx22 reviewed (commented):
I like this idea. I just quickly checked the code. And made some suggestions.

Kuhnovic commented:
Thanks for the review Glx22, I addressed your points.

I was also thinking, maybe this is a good opportunity to add a dedicated button for Many Random Industries. I've always found it a bit strange that it was just in the industry list as if it was just another type. Let me know what you guys think. 

EDIT: this is what I mean, notice the Many Random Industries entry in the list is also gone.
![image](https://user-images.githubusercontent.com/68320206/104123466-9e64f080-534b-11eb-9986-47f9dd68fb34.png)

glx22 commented:
Yeah it may be a good opportunity to remove the `many random industries` "hack" and use a proper button.
But I think other people might have a different opinion about that.

TrueBrain commented:
I agree, but I properly wouldn't put them on top of each other. Maybe next to each other is better suited? The word "industries" is a bit redundant here even, so, so "Build many random" and "Remove all" might be sufficient. Dunno, just thinking out loud :D

LC-Zorg commented:
Both changes would be very, very useful. They are sorely lacking.

> I agree, but I properly wouldn't put them on top of each other. Maybe next to each other is better suited? The word "industries" is a bit redundant here even, so, so "Build many random" and "Remove all" might be sufficient. Dunno, just thinking out loud :D

It doesn't really matter much, but I think the buttons next to each other would look a bit worse due to the asymmetry of the buttons below. Also, full sentences, due to the importance of functions and their effects, seem justified here.

Importantly, what is sorely lacking, and I would very much like to ask, is a confirmation request for both multi-enterprise construction and removal. There is nothing worse than pressing "many random" or "remove all" by accident after a long time creating a scenario ... It will be all the more important that after changing the "Remove all" button will be adjacent to the "Build".

Kuhnovic commented:
> Importantly, what is sorely lacking, and I would very much like to ask, is a confirmation request for both multi-enterprise construction and removal. 

Couldn't agree more. I pushed some additional changes. There's now a Create Random Industries button, I felt that caption was appropriate than "many random industries". Both the Remove and Create buttons ask for user confirmation.

I kept the stacked buttons since I also think it will look strange if the bottom row contains a resize button. 

![CJ6kITVlqn](https://user-images.githubusercontent.com/68320206/104239798-6056f200-545b-11eb-8e4e-7fc4fb666d66.gif)

LordAro commented:
Looks very good. Removing the "hacky" Create Random Industries "industry" option is always good

I do think the buttons would look better side-by-side though, rather than one above another

Kuhnovic commented:
I've been trying out some layouts:

![image](https://user-images.githubusercontent.com/68320206/104507485-1f8be400-55e7-11eb-8526-57802c2fb9a7.png)
**Stacked, not aligned**. Doesn't look right IMO.

![image](https://user-images.githubusercontent.com/68320206/104507617-595cea80-55e7-11eb-88cc-577fa0b821f0.png)
Resize button on separate line. This can work, but it feels a bit cluttered.

I'm working on a layout that has the resize button next to the build button, but it's difficult to line things up properly and having it look good, while also being able to switch off the scenario-editor-only buttons. This GUI system can be a bit of a pain :P

LordAro commented:
How about with create/remove buttons at the top of the window?

TrueBrain commented:
Honestly, while checking the preview (click the deployment button :D), I think on top of each other like you had it originally is totally fine (I know, I said otherwise earlier .. I changed my mind :P). Your fonts etc are making it look worse for me than it really is:

![image](https://user-images.githubusercontent.com/1663690/104511983-cd9a8c80-55ed-11eb-921e-99297ec46c25.png)

This looks perfectly fine to me :)

(well, maybe all buttons should be on top of each other, but meh, don't think that is for this PR to fix :D)

Kuhnovic commented:
> How about with create/remove buttons at the top of the window?

![image](https://user-images.githubusercontent.com/68320206/104512480-94aee780-55ee-11eb-99fa-1faad8276b16.png)
That wouldlook like this. In a way I like the separation, because now the top buttons apply to all industries, where the bottom buttons apply to the selected industry. 

(and it's a heck of a lot easier to set up)


> This looks perfectly fine to me :)

I'm also fine with keeping the first option of course ;)

LordAro commented:
It's basically a toss-up. Let's go with "top of the window, side by side"

TrueBrain commented:
![image](https://user-images.githubusercontent.com/1663690/106379421-2b7bf200-63ac-11eb-99a8-949daef6b517.png)

This is how the window now looks on a clean configuration (so no custom fonts, no font/UI zoom)

@LordAro : sure this is what you want? (honest question, but otherwise, lets merge this!)

TrueBrain commented on src/industry_gui.cpp:
```diff
@@ -610,22 +665,15 @@ class BuildIndustryWindow : public Window {
+						DoCommandP(0, this->selected_type, InteractiveRandom(), CMD_BUILD_INDUSTRY | CMD_MSG(STR_ERROR_CAN_T_CONSTRUCT_THIS_INDUSTRY));
+						this->HandleButtonClick(WID_DPI_FUND_WIDGET);
+					}
+					else {
```
Coding style: put `}` and `else {` on the same line please :)

TrueBrain commented on src/industry_gui.cpp:
```diff
@@ -578,9 +589,53 @@ class BuildIndustryWindow : public Window {
+		if (Town::GetNumItems() == 0) {
+			ShowErrorMessage(STR_ERROR_CAN_T_GENERATE_INDUSTRIES, STR_ERROR_MUST_FOUND_TOWN_FIRST, WL_INFO);
+		}
+		else {
```
Coding style: put `}` and `else {` on the same line please :)

TrueBrain commented on src/industry_gui.cpp:
```diff
@@ -578,9 +589,53 @@ class BuildIndustryWindow : public Window {
+
+		for (TileIndex tile = 0; tile < MapSize(); tile++) {
+			if (IsTileType(tile, MP_CLEAR) && GetRawClearGround(tile) == CLEAR_FIELDS) {
+				MakeClear(tile, CLEAR_GRASS, 3); // Clear farmland
```
Please put this comment before the `for`. The whole `for` is about clearing farmland, not only this `MakeClear` :)

TrueBrain commented on src/industry_gui.cpp:
```diff
@@ -392,6 +396,13 @@ class BuildIndustryWindow : public Window {
 
 		this->SetButtons();
+
+		/* Show scenario editor tools in editor */
```
sentences end with a dot. Please add a dot at the end of a sentence in a comment :)

TrueBrain reviewed (commented):
Some coding-style nitpicks :)

LC-Zorg commented:
One more approach to the layout.

Here, both buttons are below the list of enterprises, in one part with the other function buttons, what at least for me seems to be expected. Unlike previous designs, they are above the part with the explanation of the functions, what seems appropriate.
![Fund new industry window v1 0](https://user-images.githubusercontent.com/67662046/106382774-1447ff00-63c2-11eb-8fc3-0ce4b79b45e7.png)

The arrangement of buttons next to each other reduces the risk of selecting and automatically confirming the wrong function.
On the other hand, the layout with buttons one below the other allows you to minimize the size of the window if necessary. Please note that English names are one of the shorter ones. In many languages, they are much longer, which can increase the size of the window unnecessarily.

**Some examples:**
_Create many random_
_Monta satunnaista teollisuusaluetta_ - Finnish
_Grunn ghniomhachasan air thuaiream_ - Scottish Gaelic
_Daudz najausi izveidotu razatnu_ - Latvian

In Polish, the names would also be much longer, but logical abbreviations were used (removing the word "enterprises" which is not needed - just "many random"), so perhaps in other languages ​​these phrases can also be shortened.

Kuhnovic commented:
@TrueBrain nitpicking addressed ;) 

> One more approach to the layout.
> 
> Here, both buttons are below the list of enterprises, ....

I agree that it might be better to put the buttons above each other. It already looks a little funky in English tbh, and I expect it to look worse in some "lengthier" languages.

The main reason why I put the buttons above the industry list was to clearly show that they are independent of that list. Putting them below the list could create the impression that the industry selected is somehow going to affect what the buttons do, which is not the case.

TrueBrain commented:
> 
> 
> @TrueBrain nitpicking addressed ;)
> 

Did you forget to push them? :D 

Putting the buttons as pointed out in the last image makes absolutely no sense to me: `requires` and `produces` are part of the industry selected, and have nothing to do with `create random` and `remove all`. So yeah .. let's not do that for sure :P

And that settles it: at the top of the window, and on top of each other (not next to each other). It was a toss-up, but latest preview shows one clearly is sub-optimal :D

Tnx for sticking with us @Kuhnovic , and sorry we are all nitpicking like this :D I really appreciate your effort here!

Kuhnovic commented:
> > @TrueBrain nitpicking addressed ;)
> 
> Did you forget to push them? :D
> 
> Putting the buttons as pointed out in the last image makes absolutely no sense to me: `requires` and `produces` are part of the industry selected, and have nothing to do with `create random` and `remove all`. So yeah .. let's not do that for sure :P
> 
> And that settles it: at the top of the window, and on top of each other (not next to each other). It was a toss-up, but latest preview shows one clearly is sub-optimal :D
> 
> Tnx for sticking with us @Kuhnovic , and sorry we are all nitpicking like this :D I really appreciate your effort here!

Forgot to stage my changes. Still getting the hang of git, I'm a mercurial guy ;)

I can tolerate the nitpicking, better to think things through instead of going for a "fix it in the mix" later on :)

I'll swap the orientation of the buttons now and commit that as well. And I'll try not to forget staging my changes this time ;)

Eddi-z commented:
if you find yourself forgetting to stage all the time, you might use "git commit -a" instead, which automatically stages all changes. but there's very religious debates around that.

Kuhnovic commented:
> if you find yourself forgetting to stage all the time, you might use "git commit -a" instead, which automatically stages all changes. but there's very religious debates around that.

That sounds a little scary to me, but thanks for the tip!

Everything is pushed now :)

TrueBrain commented on src/widgets/industry_widget.h:
```diff
@@ -12,11 +12,14 @@
-	WID_DPI_FUND_WIDGET,    ///< Fund button.
+	WID_DPI_SCENARIO_EDITOR_PANE,            ///< Pane containing SE-only widgets.
+	WID_DPI_REMOVE_ALL_INDUSTRIES_WIDGET,    ///< Remove all industries button.
+	WID_DPI_CREATE_RANDOM_INDUSTRIES_WIDGET, ///< Remove all industries button.
```
Doing the final check, found one more thing .. should have noticed it earlier: comment doesn't match :D
